staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go

```diff
@@ -293,40 +293,46 @@ func (o *CopyOptions) copyToPod(src, dest fileSpec, options *exec.ExecOptions) e
 	return o.execute(options)
 }
 
-func (o *CopyOptions) copyFromPod(src, dest fileSpec) error {
+func (o *CopyOptions) copyFromPod(src, dest fileSpec, options *exec.ExecOptions) error {
 	if len(src.File) == 0 || len(dest.File) == 0 {
 		return errFileCannotBeEmpty
 	}
 
 	reader, outStream := io.Pipe()
-	options := &exec.ExecOptions{
-		StreamOptions: exec.StreamOptions{
-			IOStreams: genericclioptions.IOStreams{
-				In:     nil,
-				Out:    outStream,
-				ErrOut: o.Out,
-			},
 
-			Namespace: src.PodNamespace,
-			PodName:   src.PodName,
+	options.StreamOptions = exec.StreamOptions{
+		IOStreams: genericclioptions.IOStreams{
+			In:     nil,
+			Out:    outStream,
+			ErrOut: o.Out,
 		},
 
-		// TODO: Improve error messages by first testing if 'tar' is present in the container?
-		Command:  []string{"tar", "cf", "-", src.File},
-		Executor: &exec.DefaultRemoteExecutor{},
+		Namespace: src.PodNamespace,
+		PodName:   src.PodName,
 	}
+	// TODO: Improve error messages by first testing if 'tar' is present in the container?
+	options.Command = []string{"tar", "cf", "-", src.File}
 
+	errCh := make(chan error)
 	go func() {
-		defer outStream.Close()
+		defer close(errCh)
 		err := o.execute(options)
-		cmdutil.CheckErr(err)
+		// `outStream` must be Close immediately after `o.execute` returns to unblock tarReader
+		outStream.Close()
+		if err != nil {
+			errCh <- err
+		}
 	}()
 	prefix := getPrefix(src.File)
 	prefix = path.Clean(prefix)
 	// remove extraneous path shortcuts - these could occur if a path contained extra "../"
 	// and attempted to navigate beyond "/" in a remote filesystem
 	prefix = stripPathShortcuts(prefix)
-	return o.untarAll(src, reader, dest.File, prefix)
+	err := o.untarAll(src, reader, dest.File, prefix)
+	if err != nil {
+		return err
+	}
+	return <-errCh
```

Comment:
`errCh` is an unbuffered channel. Does this block if it never yields an error?

Reply:
Yes, and I think that is what we want. If source file doesn't exist in source Pod and we don't block here, the command might succeed accidentally.

Reply:
Ah I get it - you're relying on the defer at 318 to close that channel eventually.

Something might still be off though with the output.

```
 /tmp ~/Codez/kubernetes/kubernetes/_output/bin/kubectl cp nginx:/tmp/foo .
tar: Removing leading `/' from member names
tar: /tmp/foo: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
E0216 18:26:53.490999   80436 v2.go:147] io: read/write on closed pipe
command terminated with exit code 2
```